src/kernel/mutex.c

```diff
@@ -14,70 +14,168 @@
+
+    if (!prev) {
+        ci->mcs_waiting = false;
+        goto acquire;
+    }
+
+    /* prev write must occur before next */
+    ci->mcs_prev = prev;
+    write_barrier();
+    prev->mcs_next = ci;
+
+    /* limited spin to acquire lock */
     u64 spins_remain = m->spin_iterations;
     while (spins_remain-- > 0) {
-        if (mutex_cas_take(m, ctx))
-            return true;
+        if (!ci->mcs_waiting)
+            goto acquire;
+        mutex_pause();
+    }
+    mutex_debug("   spin timeout; removing from MCS list\n");
+
+    /* spin timed out; extract waiter from list and block
+
+       step A: first undo the setting of prev->mcs_next (unless we are handed
+       the lock in the meantime), thus making prev spin wait in
+       mutex_get_next() for a valid next pointer */
+    while (prev->mcs_next != ci || !compare_and_swap_64((u64*)&prev->mcs_next,
+                                                        u64_from_pointer(ci), 0)) {
+        /* unlock might hand us the lock */
+        if (!ci->mcs_waiting) {
+            compiler_barrier(); /* load aquire */
+            goto acquire;
+        }
         mutex_pause();
+
+        /* re-sample ci->mcs_prev as it may have been updated by a competing deletion */
+        prev = ci->mcs_prev;
     }
 
-    assert(!frame_is_full(ctx->frame));
+    /* step B: either wait for ci->mcs_next to be written or, if ci is at tail, rewind to prev */
+    cpuinfo next = mutex_get_next(m, ci, prev);
+    if (next) {
+        /* step C: next will spin wait on the null ci->mcs_next until we
+           update next->mcs_prev, and prev is spinning in mutex_get_next()
+           awaiting a valid next pointer - this completes the removal */
+        next->mcs_prev = prev;
+        prev->mcs_next = next;
+    }
+    ci->mcs_waiting = false;
+    on_mcs = false;
+
+    /* To cover the race where an unlock occurs between removal from MCS list
+       and insertion into waiters, attempt to grab the mutex under waiters_lock. */
+    mutex_debug("   removed; taking waiters_lock\n");
     ctx->waiting_on = m;
-    context prev_tail = pointer_from_u64(atomic_swap_64((u64*)&m->waiters_tail,
-                                                        u64_from_pointer(ctx)));
-    mutex_debug("   add ctx %p to tail, prev_tail %p\n", ctx, prev_tail);
-    if (!prev_tail) {
-        mutex_acquired(m, ctx);
+    spin_lock(&m->waiters_lock);
+    if (compare_and_swap_64((u64*)&m->turn, 0, u64_from_pointer(ctx))) {
+        ctx->waiting_on = 0;
+        mutex_debug("   mutex acquired; not suspending\n");
+        spin_unlock(&m->waiters_lock);
         return true;
     }
-    prev_tail->next_waiter = ctx;
+    mutex_debug("   inserting into waiters list and suspending\n");
+    list_insert_before(&m->waiters, &ctx->mutex_l);
+    spin_unlock(&m->waiters_lock);
     context_pre_suspend(ctx);
     context_suspend();
+  acquire:
+    while (!compare_and_swap_64((u64*)&m->turn, 0, u64_from_pointer(ctx)))
```

Comment:
I've been trying to work out a little timeout here - just counting down from m->spin_iterations - at which point this will suspend and re-schedule the context (via a new context_yield() call). I'm running into a hang which I'm trying to resolve - but I'm curious what you all think: is it even worth doing this? As far as I see it, the context needs to keep running (spinning) as it is the head contender (along with the MCS owner) - so it shouldn't just go back into the waitlist but rather be immediately scheduled. So the only advantage of breaking out of the spin here is just to make a pass through runloop before returning to this context anyway. And there's the possibility that the context could be resumed on another cpu - which supposedly could incur cache hits.

Should I even bother? Another point is that, if we enable interrupts by default during kernel execution (which was attempted last year but blocked due to issues that I think have since been resolved), external interrupts could be handled while spinning and some bh processing could occur as well. That maybe seems like better placed effort.

Reply:
I think spinning indefinitely should be avoided here, since if the mutex owner has been suspended it may not be able to resume due to the spinner context taking up all the CPU.
But it seems to me that the MCS owner doesn't necessarily have to release the MCS lock when going to the waiter list (as long as the number of MCS spin iterations cannot be infinite): it could suspend itself while holding the MCS lock, this way there will be no contention when the mutex is released, because the head of the waiter list will be the same as the MCS owner.

Reply:
I considered holding the MCS lock until the mutex unlock, and in fact originally tried to write it that way. However, for MCS to work properly, the spinning occurs on a per-CPU "local" variable - not on the mutex itself - and the previous and next links are also per-cpu variables. This is why these are stored in the cpuinfo. Even if a spin times out and the context is suspended, these per-cpu variables are still associated with _this_ mutex acquisition and would not be available for use by another.

Also, an explicit mutex handoff at the time of waking up a context means that no spinner could possibly take the lock and get work done between the point at which the context is scheduled to be resumed and the time that it wakes up and assumes ownership of the mutex. Say, for instance, that there is a considerable amount of work already queued on the runqueue, and after a mutex unlock the context return is scheduled behind it (with the mutex now owned by the scheduled context). Between the interval that this context is scheduled and resumes running, other contexts that are spinning cannot take the mutex because of this strict ordering. Removal from MCS basically trades this fair, FIFO property for (hopefully) better utilization of the resource protected by the lock.

There's yet another potential issue - not one that would affect us now, but that could become an issue as our scheduling scheme evolves (e.g. associate priorities and quanta with contexts). If we start to associate priorities with contexts and schedule returns according to priority, we could have an issue where a low-priority mutex waiter times out of its spin and suspends. An unlock wakes up the waiter, but at the same time a higher priority context becomes runnable and attempts to take the same mutex. While it is spinning, there is a priority inversion that will cause the mutex to be held up until the higher-priority context times out and joins the waiter list. This is more time that the mutex is held up without any useful work being done. It seems then that some special provision for priority inheritance would need to be taken into account.

By allowing removal from MCS, we can say that all members of the MCS list are actively-running spinners, and forward progress of ownership through them should be guaranteed provided each mutex holder behaves. I expect that, in the future, the spinning will not only check m->spin_iterations but also check if a context requires rescheduling (like need_resched() in linux - e.g. some task running as a hypothetical kernel thread exhausts its time quantum). If that is ever the case, removal from the MCS list will be essential.